lib/util/src/logging.c

```diff
@@ -1007,8 +1022,110 @@ package_source_info(struct json_context *jc)
 }
 
+static bool
+slog_tab_process_fmt(const char *fmt, char *new_fmt, size_t size)
+{
+    char *s = strchr(fmt, '%');
+
+    if (!s)
+        return false;
+
+    switch (*(++s)) {
+        case 's':
+        case 'p':
+        case 'x':
+        case 'X':
+            snprintf(new_fmt, size, "%s ", fmt);
+            return true;
+    }
+
+    snprintf(new_fmt, size, "%s ", fmt);
+    return true;
+}
+
+#define slog_tab_snprintf(pc, ...) \
+    (snprintf_append(pc->buf, pc->buf_sz, &pc->off, __VA_ARGS__))
+
+#define slog_tab_vsnprintf(pc, ...) \
+    (vsnprintf_append(pc->buf, pc->buf_sz, &pc->off, __VA_ARGS__))
+
+static
+void
+slog_tab_element_fieldv(struct slog_tab_context *pc, va_list fields)
+{
+    char *key, *fmt;
+    char  new_fmt[32];
+
+    key = va_arg(fields, char *);
+    fmt = va_arg(fields, char *);
+
+    if (!slog_tab_process_fmt(fmt, new_fmt, sizeof(new_fmt)))
+        return;
+
+    slog_tab_snprintf(pc, "%s ", key);
+    slog_tab_vsnprintf(pc, new_fmt, fields);
+}
+
+static void
+slog_tab_format_payload(struct slog_tab_context *pc, va_list payload)
+{
+    int   token, cnt;
+    char *key, *fmt;
+    void *val;
+
+    token = va_arg(payload, int);
+
+    while (token) {
+        switch (token) {
+            case _SLOG_START_TOKEN:
+                slog_tab_snprintf(pc, "%s ", HSE_MARK);
+                slog_tab_element_fieldv(pc, payload);
+                break;
+            case _SLOG_CHILD_START_TOKEN:
+                key = va_arg(payload, char *);
+                slog_tab_snprintf(pc, "%s", key);
+                break;
+            case _SLOG_FIELD_TOKEN:
+                slog_tab_element_fieldv(pc, payload);
+                break;
+            case _SLOG_LIST_TOKEN:
+                key = va_arg(payload, char *);
+                fmt = va_arg(payload, char *);
+                cnt = va_arg(payload, int);
+                val = va_arg(payload, void *);
+                slog_tab_snprintf(pc, "%s[", key);
+                for (int i = 0; i < cnt; i++) {
+                    switch (fmt[1]) {
+                        case 's':
+                            slog_tab_snprintf(pc, "%s,", *((char **)val + i));
+                            break;
+                        case 'd':
+                            slog_tab_snprintf(pc, "%d,", *((int *)val + i));
+                            break;
+                        case 'l':
+                            slog_tab_snprintf(pc, "%lu,", *((unsigned long *)val + i));
+                            break;
+                        default:
+                            i=cnt;
+                            break;
+                    }
+                }
+                slog_tab_snprintf(pc, "]");
+                break;
+            case _SLOG_CHILD_END_TOKEN:
+                break;
+            case _SLOG_END_TOKEN:
+                break;
+            default:
+                break;
+        }
+
+        token = va_arg(payload, int);
+    }
+}
+
```

Comment:
Identifiers starting with an underscore and followed by a capital letter are reserved by for use by the implementation.

Also, case label indent...

Reply:
I'll push a follow-on PR with that and some other symbol renames:
```
   s/_SLOG_/SLOG_TOK_/
   s/HSE_SLOG/SLOG/
```

Reply:
case label fix coming will be added to this PR.

Reply:
Very cool. Thanks for fixing naming